packages/contract/src/DaimoUSDCSwapper.sol

```diff
@@ -0,0 +1,302 @@
+        // Can trace to debug oracle issues if needed.
+      }
+    }
+  }
+
+  // Direct 1-hop quote: [tokenIn -> tokenOut]
+  function quoteDirect(
+    uint128 amountIn,
+    IERC20 tokenIn,
+    IERC20 tokenOut
+  ) public view returns (uint256 amountOut, uint24 fee) {
+    int24 tick;
+    address swapPool;
+    (swapPool, tick, fee) = getBestPoolTick(tokenIn, tokenOut);
+
+    if (swapPool == address(0)) return (0, 0);
+
+    amountOut = OracleLibrary.getQuoteAtTick({
+      tick: tick,
+      baseAmount: amountIn,
+      baseToken: address(tokenIn),
+      quoteToken: address(tokenOut)
+    });
+  }
+
+  // 2-hop paths: [tokenIn -> hopToken -> tokenOut]
+  function quoteViaHop(
+    uint128 amountIn,
+    IERC20 tokenIn,
+    IERC20 tokenOut
+  ) public view returns (uint256 amountOut, bytes memory swapPath) {
+    for (uint256 i = 0; i < hopTokens.length; i++) {
+      IERC20 hopToken = hopTokens[i];
+
+      if (hopToken == tokenIn) continue; // Covered by direct quote
+      if (hopToken == tokenOut) continue; // Covered by direct quote
+
+      (address poolOne, int24 tick, uint24 feeOne) = getBestPoolTick(
+        tokenIn,
+        hopToken
+      );
+
+      if (poolOne == address(0)) continue;
+
+      uint256 hopAmountOut = OracleLibrary.getQuoteAtTick({
+        tick: tick,
+        baseAmount: amountIn,
+        baseToken: address(tokenIn),
+        quoteToken: address(hopToken)
+      });
+
+      if (hopAmountOut > _MAX_UINT128) continue;
+
+      (uint256 pathAmountOut, uint24 feeTwo) = quoteDirect(
+        uint128(hopAmountOut),
+        hopToken,
+        tokenOut
+      );
+
+      if (pathAmountOut > amountOut) {
+        amountOut = pathAmountOut;
+        swapPath = abi.encodePacked(
+          address(tokenIn),
+          feeOne,
+          address(hopToken),
+          feeTwo,
+          address(tokenOut)
+        );
+      }
+    }
+  }
+
+  // Fetch a best-effort quote for a given exact input token pair.
+  // token = 0x0 refers to ETH.
+  function quote(
+    uint128 amountIn,
+    IERC20 tokenIn,
+    IERC20 tokenOut
+  ) public view returns (uint256 amountOut, bytes memory swapPath) {
+    if (address(tokenIn) == address(0)) tokenIn = weth;
+    if (address(tokenOut) == address(0)) tokenOut = weth;
+
+    // Same token swap.
+    if (tokenIn == tokenOut) {
+      amountOut = amountIn;
+      swapPath = new bytes(0);
+      return (amountOut, swapPath);
+    }
+    (uint256 directAmountOut, uint24 directFee) = quoteDirect(
+      amountIn,
+      tokenIn,
+      tokenOut
+    );
+
+    (uint256 hopAmountOut, bytes memory swapPathHop) = quoteViaHop(
+      amountIn,
+      tokenIn,
+      tokenOut
+    );
+
+    if (directAmountOut > hopAmountOut) {
+      amountOut = directAmountOut;
+      swapPath = abi.encodePacked(
+        address(tokenIn),
+        directFee,
+        address(tokenOut)
+      );
+    } else {
+      amountOut = hopAmountOut;
+      swapPath = swapPathHop;
+    }
+  }
```

Comment:
yesterday, we hotfixed API to pull `quoteDirect` instead of `quote` for a somewhat subtle reason.

our current quoter works as follows:
- uses the best liquidity pool *per pair* for a given path
- uses the highest price *per path* across (direct path + a few 2-pool paths)

problem is, what if the highest price is ~fake news? we saw an example yesterday.

if there is a hop path ~no liquidity (=best liquidity for at least one of the pairs along that route is very low) but highest price, we quote that path.

--

here's an idea for fixing this.
- `quote` > `quoteFromOracle` and just return price, no path. this is a rough estimate of best-available price, gas efficient, & robust to manipulation using TWAP/TWAL.

- a new function, `quoteBestPath`

use https://docs.uniswap.org/contracts/v3/reference/periphery/interfaces/IQuoterV2 to get the real amountOut for direct path vs best hop path and return that